### ISSUE TITLE
Recursion detected does not make sense

### ISSUE THREAD

**Issue by PaulHeisterkamp** [microsoft/Dynamics365FO-AppChecker#20]
Recursion detected: I think this rule do not make sense because a modern programming language it should be possible to use recursions. The argument for banning recursions is "...we had recursion related problems in production environments to often...". For me this is also no agreement for putting the recursions on a "black list". I think it would make more sense to add runtime checks to restrict the recursion depth.

**Comment by PalleAgermark** [comment]
I agree. If you need recursion, you'll now have to implement solutions "clever" enough to go under the radar of the App Checker. Probably with even more interesting recursion bugs as a result. Recursion requirements wont go away. Maybe Microsoft could give us some sort of framework to use to ensure that endless recursions are avoidable.

**Comment by dickwenning** [comment]
In my old dayes (1990's) on Telefoon switches software we also banned recursion,
the risk of non ending loop is there with adition high CPU usages.

**Comment by AndreasHassing** [comment]
This discussion is interesting 😊, thanks for bringing it up.

@PaulHeisterkamp: how would you detect runaway recursions? We already have some protection we can toggle on for environments that get stack overflow crashes a lot, but it comes with a performance penalty. Should all environments pay that price so developers can use recursions without warning?

@PalleAgermark: what kind of framework? The only thing that comes to my mind is what F# does with the `rec` keyword, but that might not work for us since the order of method declarations does not matter in X++ - so how can we know for sure method A does not call method B somewhere down the line (when method B calls method A).

**Comment by PaulHeisterkamp** [comment]
@AndreasHassing I would vote for the [Recursive] attribute to reduce the effort for checking recursion depth. In addition to that I would limit to a "static" number e. g. 100
